packages/techdocs-common/src/stages/prepare/preparers.ts

```diff
@@ -13,14 +13,45 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
-import { PreparerBase, RemoteProtocol, PreparerBuilder } from './types';
+import { Logger } from 'winston';
+import { UrlReader } from '@backstage/backend-common';
 import { Entity } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { DirectoryPreparer, CommonGitPreparer, UrlPreparer } from '.';
+import { PreparerBase, RemoteProtocol, PreparerBuilder } from './types';
 import { parseReferenceAnnotation } from '../../helpers';
 
+type factoryOptions = {
+  logger: Logger;
+  reader: UrlReader;
+};
+
 export class Preparers implements PreparerBuilder {
   private preparerMap = new Map<RemoteProtocol, PreparerBase>();
 
+  static async fromConfig(
+    // @ts-ignore
+    // Config not used now, but will be used in urlPreparer when it starts using
+    // @backstage/integration to get the tokens for providers.
+    config: Config,
+    { logger, reader }: factoryOptions,
+  ): Promise<PreparerBuilder> {
+    const preparers = new Preparers();
+
+    const directoryPreparer = new DirectoryPreparer(logger);
+    preparers.register('dir', directoryPreparer);
+
+    const commonGitPreparer = new CommonGitPreparer(logger);
+    preparers.register('github', commonGitPreparer);
+    preparers.register('gitlab', commonGitPreparer);
+    preparers.register('azure/api', commonGitPreparer);
```

Comment:
Not sure how it's done elsewhere, but i feel that it's sort of bad practice to add all preparers. Can we add just the ones we've specified we want in config? Feels like most setups would only want to use one preparer, making all others nothing but potential security issues.

Reply:
Interesting thought. We could discuss this in context of both Scaffolder and TechDocs.

https://github.com/backstage/backstage/blob/ebbc133eaa0a4883db97572cca4b1c94f680d95e/plugins/scaffolder-backend/src/scaffolder/stages/prepare/preparers.ts#L73-L83

In my opinion, the backend plugins should be able to support most source code hosting providers. But adding another layer of "allowed list" for these types of providers is a possibility for Scaffolder/TechDocs.

Similar to how the CSP config works for loading images, for example. We technically supports all sources, but there is an [allowed list](https://github.com/backstage/backstage/blob/ebbc133eaa0a4883db97572cca4b1c94f680d95e/packages/backend-common/src/service/lib/ServiceBuilderImpl.ts#L53).

Reply:
I guess if the scaffolder does it that way we're good. It also makes sense to do things in a similar way to make it easier to get started.

Reply:
The catalog backend has a similar but somewhat different approach that might be worthwhile looking at as well https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/service/CatalogBuilder.ts